app/services/map_activity_service.rb

```diff
@@ -0,0 +1,94 @@
+class MapActivityService
+
+  def self.subject_line(map)
+    'Activity on map ' + map.name
+  end
+
+  def self.summarize_data(map, user, until_moment = DateTime.now)
+    results = {
+      stats: {}
+    }
+
+    since = until_moment - 24.hours
+
+    message_count = Message.where(resource: map)
+                           .where("created_at > ? AND created_at < ?", since, until_moment)
+                           .where.not(user: user).count
+    if message_count > 0
+      results[:stats][:messages_sent] = message_count
+    end
+
+    moved_count = Event.where(kind: 'topic_moved_on_map', map: map)
+                       .where("created_at > ? AND created_at < ?", since, until_moment)
+                       .where.not(user: user).group(:eventable_id).count
+    if moved_count.keys.length > 0
+      results[:stats][:topics_moved] = moved_count.keys.length
+    end
+
+    topics_added_events = Event.where(kind: 'topic_added_to_map', map: map)
+                               .where("created_at > ? AND created_at < ?", since, until_moment)
+                               .where.not(user: user)
+                               .order(:created_at)
+
+    topics_removed_events = Event.where(kind: 'topic_removed_from_map', map: map)
+                                 .where("created_at > ? AND created_at < ?", since, until_moment)
+                                 .where.not(user: user)
+                                 .order(:created_at)
+
+    topics_added_to_include = {}
+    topics_added_events.each do |ta|
+      num_adds = topics_added_events.where(eventable_id: ta.eventable_id).count
+      num_removes = topics_removed_events.where(eventable_id: ta.eventable_id).count
+      topics_added_to_include[ta.eventable_id] = ta if num_adds > num_removes
+    end
+    if topics_added_to_include.keys.length > 0
+      results[:stats][:topics_added] = topics_added_to_include.keys.length
+      results[:topics_added] = topics_added_to_include.values
+    end
+
+    topics_removed_to_include = {}
+    topics_removed_events.each do |ta|
+      num_adds = topics_added_events.where(eventable_id: ta.eventable_id).count
+      num_removes = topics_removed_events.where(eventable_id: ta.eventable_id).count
+      topics_removed_to_include[ta.eventable_id] = ta if num_removes > num_adds
+    end
+    if topics_removed_to_include.keys.length > 0
+      results[:stats][:topics_removed] = topics_removed_to_include.keys.length
+      results[:topics_removed] = topics_removed_to_include.values
+    end
+
+    synapses_added_events = Event.where(kind: 'synapse_added_to_map', map: map)
```

Comment:
Realized I need to get down to granular permissioning here, related to topics and synapses because stuff can be going on secretly on a map and other people shouldn't know.